File: python/fit/RooFitter.py

#
#   @author: Omar Moreno <omoreno1@ucsc.edu>
#
#

import ROOT as r

class RooFitter(object):

    def __init__(self, root_file_name, histogram_name) : 
        self.root_file = r.TFile(root_file_name)
        self.histogram = self.root_file.Get(histogram_name)

    def fit_to_gaussian(self) :
        histogram_min = self.histogram.GetXaxis().GetXmin()
        histogram_max = self.histogram.GetXaxis().GetXmax()
        x = r.RooRealVar("x", "x", histogram_min, histogram_max)
        plot = x.frame()
        arg_list = r.RooArgList(x)
        histogram_data = r.RooDataHist("histogram_data", "histogram_data", arg_list, self.histogram)
        mean = histogram_data.mean(x)
        sigma = histogram_data.sigma(x)
        print "Histogram mean: " + str(mean)
        print "Histogram sigma: " + str(sigma)

        mean_var = RooRealVar("mean_var", "Gaussian - Mean", mean, mean - 4*sigma, mean + 4*sigma)
        sigma_var = RooRealVar("sigma_var", "Gaussian - Sigma", sigma, 0, 10*sigma)
        
        gaussian = RooGaussian("gaussian", "Gaussian", x, mean_var, sigma_var)
        fit_result = gaussian.fitTo(histogram_data)
        histogram_data.plotOn(plot)
        gaussian.plotOn(plot)

        plot.Draw()

    def fit_to_crystal_ball(self) : 

        histogram_min = self.histogram.GetXaxis().GetXmin()
        histogram_max = self.histogram.GetXaxis().GetXmax()

        x = r.RooRealVar("x", "x", histogram_min, histogram_max)
        plot = x.frame()

        arg_list = r.RooArgList(x)
        histogram_data = r.RooDataHist("histogram_data", "histogram_data", arg_list, self.histogram)

        mean = histogram_data.mean(x)
        sigma = histogram_data.sigma(x)
        
        mean_var = r.RooRealVar("#mu", "mean", mean, mean - 10*sigma, mean + 10*sigma)
        sigma_var = r.RooRealVar("#sigma", "mean", sigma, 0, 10*sigma)
        alpha_var = r.RooRealVar("#alpha", "alpha", -2, -100, -1)
        n_var = r.RooRealVar("n", "n", 0, 0, 20)

        cb = r.RooCBShape("cb", "cb", x, mean_var, sigma_var, alpha_var, n_var)

        cb.fitTo(histogram_data)

        histogram_data.plotOn(plot)
        cb.plotOn(plot)
        cb.paramOn(plot)

        plot.Draw()

File: src/utils/RooFitter.cxx
/**
 * @file RooFitter.cxx
 * @brief A common set of fits done using RooFit
 * @author Omar Moreno <omoreno1@ucsc.edu> \n
 *         Santa Cruz Institute for Particle Physics
 *         University of California, Santa Cruz
 * @date August 24, 2015
 */

#include <RooFitter.h>

using namespace RooFit; 

RooPlot* RooFitter::fitToGaussian(TH1* histogram, RooRealVar var) { 

    RooDataHist histogram_data("data", histogram->GetName(), var, histogram); 

    double mean = histogram->GetMean(); 
    double rms = histogram->GetRMS(); 

    RooRealVar gauss_mean("gauss_mean", "mean", mean, mean - 4*rms, mean + 4*rms);  
    RooRealVar gauss_sigma("gauss_sigma", "#sigma", rms, rms - 3*rms, rms + 3*rms); 
    RooGaussian gaussian("gaussian", "gaussian", var, gauss_mean, gauss_sigma); 

    RooFitResult* result = gaussian.fitTo(histogram_data, Range(mean - 2.*rms, mean + 2.*rms), Save());

    RooPlot* plot = var.frame(); 
    histogram_data.plotOn(plot); 
    gaussian.plotOn(plot, Range(0, 5.0));
    gaussian.paramOn(plot);   

    return plot; 
}

RooPlot* RooFitter::fitToDoubleGaussian(TH1* histogram, RooRealVar var) { 
    
    RooDataHist histogram_data("data", histogram->GetName(), var, histogram); 

    double mean = histogram->GetMean(); 
    double rms = histogram->GetRMS(); 

    RooRealVar sig_gauss_mean("sig_gauss_mean", "mean", mean, mean - 3*rms, mean + 3*rms);  
    RooRealVar sig_gauss_sigma("sig_gauss_sigma", "#sigma", rms, rms - 3*rms, rms + 3*rms); 
    RooGaussian sig_gaussian("sig_gaussian", "gaussian", var, sig_gauss_mean, sig_gauss_sigma); 

    RooRealVar bkg_gauss_mean("bkg_gauss_mean", "mean", mean, mean - 10*rms, mean + 10*rms);  
    RooRealVar bkg_gauss_sigma("bkg_gauss_sigma", "#sigma", rms, rms - 10*rms, rms + 10*rms); 
    RooGaussian bkg_gaussian("bkg_gaussian", "gaussian", var, bkg_gauss_mean, bkg_gauss_sigma); 

    RooRealVar n_sig("n_sig", "signal events", 5000, 0, 100000);
    RooRealVar n_bkg("n_bkg", "background events", 500, 0, 10000); 
    RooAddPdf sum("sum", "gauss+gauss", RooArgList(sig_gaussian, bkg_gaussian), RooArgList(n_sig, n_bkg)); 
    
    RooFitResult* result = sum.fitTo(histogram_data, Save()); 
    
    RooPlot* plot = var.frame(); 
    histogram_data.plotOn(plot); 
    sig_gaussian.plotOn(plot, LineStyle(kDashed));
    bkg_gaussian.plotOn(plot, LineStyle(kDashed), LineColor(kRed)); 
    sum.plotOn(plot);
    sum.paramOn(plot); 

    return plot;  
}
